Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -0,0 +1,995 @@
+local macro "C_simp" : tactic =>
+  `(tactic| simp only [C_0, C_1, C_neg, C_add, C_sub, C_mul, C_pow])
+
+open scoped Polynomial PolynomialPolynomial
+
+/-- The polynomial $W(X, Y) := Y^2 + a_1XY + a_3Y - (X^3 + a_2X^2 + a_4X + a_6)$ associated to a
+Weierstrass curve `W` over `R`. For ease of polynomial manipulation, this is represented as a term
+of type `R[X][X]`, where the inner variable represents $X$ and the outer variable represents $Y$.
+For clarity, the alternative notations `Y` and `R[X][Y]` are provided in the `PolynomialPolynomial`
+scope to represent the outer variable and the bivariate polynomial ring `R[X][X]` respectively. -/
+protected noncomputable def polynomial : R[X][Y] :=
+  Y ^ 2 + C (C W.a₁ * X + C W.a₃) * Y - C (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)
+#align weierstrass_curve.polynomial WeierstrassCurve.polynomial
+
+lemma polynomial_eq :
+    W.polynomial =
+      Cubic.toPoly
+        ⟨0, 1, Cubic.toPoly ⟨0, 0, W.a₁, W.a₃⟩, Cubic.toPoly ⟨-1, -W.a₂, -W.a₄, -W.a₆⟩⟩ := by
+  simp only [WeierstrassCurve.polynomial, Cubic.toPoly]
+  C_simp
+  ring1
+#align weierstrass_curve.polynomial_eq WeierstrassCurve.polynomial_eq
+
+lemma polynomial_ne_zero [Nontrivial R] : W.polynomial ≠ 0 := by
+  rw [polynomial_eq]
+  exact Cubic.ne_zero_of_b_ne_zero one_ne_zero
+#align weierstrass_curve.polynomial_ne_zero WeierstrassCurve.polynomial_ne_zero
+
+-- porting note: removed `@[simp]`
+lemma degree_polynomial [Nontrivial R] : W.polynomial.degree = 2 := by
+  rw [polynomial_eq]
+  exact Cubic.degree_of_b_ne_zero' one_ne_zero
+#align weierstrass_curve.degree_polynomial WeierstrassCurve.degree_polynomial
+
+-- porting note: removed `@[simp]`
+lemma natDegree_polynomial [Nontrivial R] : W.polynomial.natDegree = 2 := by
+  rw [polynomial_eq]
+  exact Cubic.natDegree_of_b_ne_zero' one_ne_zero
+#align weierstrass_curve.nat_degree_polynomial WeierstrassCurve.natDegree_polynomial
+
+lemma monic_polynomial : W.polynomial.Monic := by
+  nontriviality R
+  simpa only [polynomial_eq] using Cubic.monic_of_b_eq_one'
+#align weierstrass_curve.monic_polynomial WeierstrassCurve.monic_polynomial
+
+-- porting note: replaced two `any_goals` proofs with two pairs of repeated proofs
+lemma irreducible_polynomial [IsDomain R] : Irreducible W.polynomial := by
+  by_contra h
+  rcases (W.monic_polynomial.not_irreducible_iff_exists_add_mul_eq_coeff W.natDegree_polynomial).mp
+    h with ⟨f, g, h0, h1⟩
+  simp only [polynomial_eq, Cubic.coeff_eq_c, Cubic.coeff_eq_d] at h0 h1
+  apply_fun degree at h0 h1
+  rw [Cubic.degree_of_a_ne_zero' <| neg_ne_zero.mpr <| one_ne_zero' R, degree_mul] at h0
+  apply (h1.symm.le.trans Cubic.degree_of_b_eq_zero').not_lt
+  rcases Nat.WithBot.add_eq_three_iff.mp h0.symm with h | h | h | h
+  · rw [degree_add_eq_right_of_degree_lt] <;> simp only [h]
+  · rw [degree_add_eq_right_of_degree_lt] <;> simp only [h]
+  · rw [degree_add_eq_left_of_degree_lt] <;> simp only [h]
+  · rw [degree_add_eq_left_of_degree_lt] <;> simp only [h]
```

Comment:
```suggestion
  iterate 2
    rw [degree_add_eq_right_of_degree_lt] <;> simp only [h]
  iterate 2
    rw [degree_add_eq_left_of_degree_lt] <;> simp only [h]
```

Reply:
Also, it's easier if you put the porting notes right before the affected part of the proof, rather than before the declaration.